File: python/321.py
from typing import List


class Solution:
    def maxNumber(self, nums1: List[int], nums2: List[int], k: int) -> List[int]:
        n, m = len(nums1), len(nums2)
        nums1_right = [n] * n
        stack = []
        for i in range(n - 1, -1, -1):
            while stack and nums1[stack[-1]] <= nums1[i]:
                stack.pop()
            if stack:
                nums1_right[i] = stack[-1]
            stack.append(i)

        nums2_right = [m] * m
        stack = []
        for i in range(m - 1, -1, -1):
            while stack and nums2[stack[-1]] <= nums2[i]:
                stack.pop()
            if stack:
                nums2_right[i] = stack[-1]
            stack.append(i)

        print("nums1", nums1)
        print("nums2", nums2)
        print("=" * 20)
        print(nums1_right)
        print(nums2_right)
        print("=" * 20)

        res = []
        i = j = 0
        while len(res) < k:
            if i < n and nums1[i] == 9:
                res.append(nums1[i])
                i += 1
            elif j < m and nums2[j] == 9:
                res.append(nums2[j])
                j += 1
            elif i < n and j < m:
                if nums1[i] == nums2[j]:
                    pass
                elif nums1[i] < nums2[j]:
                    if (n + m - nums1_right[i] - j) >= k - len(res) and nums1_right[i] < n:
                        i = nums1_right[i]
                    else:
                        res.append(nums2[j])
                        j += 1
                else:
                    if (n + m - i - nums2_right[j]) >= k - len(res) and nums2_right[j] < m:
                        j = nums2_right[j]
                    else:
                        res.append(nums1[i])
                        i += 1
            elif i < n:
                if (n - nums1_right[i]) >= k - len(res) and nums1_right[i] < n:
                    i = nums1_right[i]
                else:
                    res.append(nums1[i])
                    i += 1
            elif j < m:
                if (m - nums2_right[j]) >= k - len(res) and nums2_right[j] < m:
                    j = nums2_right[j]
                else:
                    res.append(nums2[j])
                    j += 1
        print(res)


def main():
    # nums1 = [3, 4, 6, 5]
    # nums2 = [9, 1, 2, 5, 8, 3]
    # k = 5
    nums1 = [6, 7]
    nums2 = [6, 0, 4]
    k = 5
    Solution().maxNumber(nums1, nums2, k)


if __name__ == "__main__":
    main()

File: go/problems/321.go
package problems

import "slices"

func maxNumber321(nums1 []int, nums2 []int, k int) []int {
	res := make([]int, k)
	if len(nums1) > len(nums2) {
		nums1, nums2 = nums2, nums1
	}

	maxNum := func(size int, nums []int) []int {
		stack := []int{}
		for i, digit := range nums {
			for len(stack) > size-min(size, len(nums)-i) && digit > stack[len(stack)-1] {
				stack = stack[:len(stack)-1]
			}

			if len(stack) < size {
				stack = append(stack, digit)
			}
		}
		return stack
	}

	merge := func(n1, n2 []int) []int {
		if len(n1) == 0 {
			return n2
		}
		if len(n2) == 0 {
			return n1
		}

		res := make([]int, 0, len(n1)+len(n2))

		p1, p2 := 0, 0
		for p1 < len(n1) && p2 < len(n2) {
			if n1[p1] > n2[p2] {
				res = append(res, n1[p1])
				p1++
				continue
			}

			if n1[p1] < n2[p2] {
				res = append(res, n2[p2])
				p2++
				continue
			}

			i, j := p1, p2
			for i < len(n1) && j < len(n2) && n1[i] == n2[j] {
				i++
				j++
			}

			if i == len(n1) || (j < len(n2) && n1[i] < n2[j]) {
				res = append(res, n2[p2])
				p2++
			}

			if j == len(n2) || (i < len(n1) && n1[i] > n2[j]) {
				res = append(res, n1[p1])
				p1++
			}

		}

		res = append(res, n1[p1:]...)
		res = append(res, n2[p2:]...)

		return res
	}

	for m := range min(len(nums1), k) + 1 {
		if k-m <= len(nums2) {
			n1 := maxNum(m, nums1)
			n2 := maxNum(k-m, nums2)
			num := merge(n1, n2)
			if slices.Compare(num, res) > 0 {
				res = num
			}
		}
	}

	return res

}
